### PR TITLE
feat: wrap FindexCloud in Pyo3

### PR DESCRIPTION
**Pull request by ackRow**

### CODE REVIEW

File: crates/findex/Cargo.toml [hunk 1]
```diff
@@ -15,7 +15,7 @@ name = "cloudproof_findex"
 [features]
 cloud = ["serde", "reqwest"]
 ffi = ["serde", "serde_json", "cloud", "cosmian_ffi", "tokio"]
-python = ["pyo3"]
+python = ["pyo3", "cloud", "tokio"]
```
Comment (tbrezot) [inline crates/findex/Cargo.toml:18]: Do you want to ship "cloud" all the time?

File: crates/findex/src/pyo3/py_api.rs [hunk 2]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+#[pyclass]
+pub struct FindexCloud(FindexCloudRust);
+
+#[pymethods]
+impl FindexCloud {
+    #[new]
+    pub fn new(token: &str, base_url: Option<String>) -> PyResult<Self> {
+        Ok(Self(pyo3_unwrap!(
+            FindexCloudRust::new(token, base_url),
+            "error reading token"
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:531]: Can the error only happen on token read?

File: crates/findex/src/pyo3/py_api.rs [hunk 3]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+        )))
+    }
+
+    /// Upserts the given relations between `IndexedValue` and `Keyword` into
+    /// Findex tables. After upserting, any search for a `Word` given in the
+    /// aforementioned relations will result in finding (at least) the
+    /// corresponding `IndexedValue`.
+    ///
+    /// Parameters
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:540]: Is there a reason for not using `# Parameters`?

File: crates/findex/src/pyo3/py_api.rs [hunk 4]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+    ///
+    /// - `indexed_values_and_keywords` : map of `IndexedValue` to keywords
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:542]:
```suggestion
    /// - `indexed_values_and_keywords` : map of `IndexedValue` to `Keyword`
```

File: crates/findex/src/pyo3/py_api.rs [hunk 5]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+    /// - `label`                   : public label used in keyword hashing
+    /// - `max_results_per_keyword` : maximum number of results to fetch per
+    ///   keyword
+    /// - `max_depth`               : maximum recursion level allowed
+    /// - `fetch_chains_batch_size` : batch size during fetch chain
+    /// - `progress_callback`       : optional callback to process intermediate
+    ///   search results.
+    ///
+    /// Returns: List[IndexedValue]
+    // use `u32::MAX` for `max_result_per_keyword`
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:582]: Maybe we don't need this? If you still need it, maybe you can place it above the specific line (`max_result_per_keyword = ...`)

```suggestion
```

File: crates/findex/src/pyo3/py_api.rs [hunk 6]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+        fetch_chains_batch_size: usize,
+        py: Python,
+    ) -> PyResult<HashMap<String, Vec<PyObject>>> {
+        let keywords_set: HashSet<Keyword> = keywords
+            .iter()
+            .map(|keyword| Keyword::from(*keyword))
+            .collect();
+
+        // We want to forward error code returned by callbacks to the parent caller to
+        // do error management client side.
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:608]: I don't understand what that means.

File: crates/findex/src/pyo3/py_api.rs [hunk 7]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+    }
+
+    /// Generate a new Findex Cloud token with reduced permissions
+    #[staticmethod]
+    pub fn derive_new_token(token: String, search: bool, index: bool) -> PyResult<String> {
+        let mut token = pyo3_unwrap!(Token::from_str(&token), "error reading token");
+
+        pyo3_unwrap!(
+            token.reduce_permissions(search, index),
+            "error reducing token's permissions"
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:640]:
```suggestion
            "error reducing token permissions"
```

File: crates/findex/src/pyo3/py_api.rs [hunk 8]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+                    "error reading seed from Uint8Slice"
+                ),
+                pyo3_unwrap!(
+                    uint8slice_to_seed(upsert_entries_seed, "upsert_entries_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
+                pyo3_unwrap!(
+                    uint8slice_to_seed(insert_chains_seed, "insert_chains_seed"),
+                    "error reading seed from Uint8Slice"
+                ),
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:673]: Change the return type of `uintslice_to_seed` to return a pyo3 error

File: crates/findex/src/pyo3/py_api.rs [hunk 9]
```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
...
+    py: Python,
+) -> HashMap<String, Vec<PyObject>> {
+    search_results
+        .iter()
+        .map(|(keyword, locations)| {
+            (
+                // Convert keyword to string or base64
+                match keyword.clone().try_into_string() {
+                    Ok(s) => s,
+                    Err(_) => format!("{keyword:?}"),
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:720]: I think the debug does not convert to base64 anymore, you should convert manually here.

File: crates/findex/src/pyo3/py_api.rs (hunk 3 above)
Comment (ackRow) [reply crates/findex/src/pyo3/py_api.rs:540]: I used the same formatting as the previous functions

File: crates/findex/Cargo.toml (hunk 1 above)
Comment (ackRow) [reply crates/findex/Cargo.toml:18]: I think it's more convenient this way.
Plus it's the same as the ffi and wasm.

File: crates/findex/python/cloudproof_findex/__init__.py [hunk 10]
```diff
@@ -1,6 +1,12 @@
 # -*- coding: utf-8 -*-
+from typing import Dict, List, Sequence, Union
+
 from .cloudproof_findex import *
 
+IndexedValuesAndKeywords = Dict[Union[Location, Keyword], Sequence[Union[str, Keyword]]]
```
Comment (tbrezot) [inline crates/findex/python/cloudproof_findex/__init__.py:6]: For my understanding, why not using a `List` here like for `SearchResults`?

File: crates/findex/src/pyo3/mod.rs [hunk 11]
```diff
@@ -10,14 +9,119 @@ macro_rules! pyo3_unwrap {
     };
 }
 
+/// Implements the basic functionalities of keyword and location in python.
+///
+/// # Parameters
+///
+/// - `type_name`   : name of the key type
+macro_rules! impl_python_byte {
+    ($py_type:ty, $rust_type:ty, $name:tt) => {
```
Comment (tbrezot) [inline crates/findex/src/pyo3/mod.rs:18]: Would a literal be okay here?
```suggestion
    ($py_type:ty, $rust_type:ty, $name:literal) => {
```

File: crates/findex/src/pyo3/mod.rs [hunk 12]
```diff
@@ -10,14 +9,119 @@ macro_rules! pyo3_unwrap {
     };
 }
 
+/// Implements the basic functionalities of keyword and location in python.
+///
+/// # Parameters
+///
+/// - `type_name`   : name of the key type
```
Comment (tbrezot) [inline crates/findex/src/pyo3/mod.rs:16]: Doc does not match macros signature

File: crates/findex/src/pyo3/mod.rs [hunk 13]
```diff
@@ -10,14 +9,119 @@ macro_rules! pyo3_unwrap {
...
+
+            /// Converts to bytes.
+            fn __bytes__(&self, py: Python) -> Py<PyBytes> {
+                PyBytes::new(py, &self.0).into()
+            }
+
+            /// Converts to int.
+            fn __int__(&self) -> PyResult<i64> {
+                let slice: &[u8] = &self.0;
+                let arr: [u8; 8] = slice.try_into()?;
```
Comment (tbrezot) [inline crates/findex/src/pyo3/mod.rs:55]:
```suggestion
                let arr: [u8; 8] = self.0.try_into()?;
```

File: crates/findex/src/pyo3/mod.rs [hunk 14]
```diff
@@ -10,14 +9,119 @@ macro_rules! pyo3_unwrap {
...
+                }
+            }
+
+            /// Converts to bytes.
+            fn __bytes__(&self, py: Python) -> Py<PyBytes> {
+                PyBytes::new(py, &self.0).into()
+            }
+
+            /// Converts to int.
+            fn __int__(&self) -> PyResult<i64> {
```
Comment (tbrezot) [inline crates/findex/src/pyo3/mod.rs:53]: Why do you need to be able to convert into a i64?

I would rather let the user transform bytes into the specific type in use.

File: crates/findex/src/pyo3/py_api.rs [hunk 15]
```diff
@@ -463,24 +160,21 @@ impl InternalFindex {
...
             Some(callback) => callback,
             None => self.default_progress_callback.clone(),
         };
 
-        let keywords_set: HashSet<Keyword> = keywords
-            .iter()
-            .map(|keyword| Keyword::from(*keyword))
-            .collect();
+        let keywords_set: HashSet<Keyword> =
+            keywords.iter().map(|keyword| keyword.0.clone()).collect();
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:177]: Needless copy (you already have the ownership of the keywords):
```suggestion
        let keywords_set: HashSet<Keyword> =
            keywords.into_iter().map(|keyword| keyword.0).collect();
```

File: crates/findex/src/pyo3/py_api.rs [hunk 16]
```diff
@@ -554,3 +232,188 @@ impl InternalFindex {
...
+    /// Upserts the given relations between `IndexedValue` and `Keyword` into
+    /// Findex tables. After upserting, any search for a `Word` given in the
+    /// aforementioned relations will result in finding (at least) the
+    /// corresponding `IndexedValue`.
+    ///
+    /// Parameters
+    ///
+    /// - `indexed_values_and_keywords` : map of `IndexedValue` to `Keyword`
+    /// - `token`                       : Findex token
+    /// - `label`                       : label used to allow versioning
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:250]: Missing parameter in documentation: review all your documentations

File: crates/findex/src/pyo3/py_api.rs [hunk 17]
```diff
@@ -554,3 +232,188 @@ impl InternalFindex {
...
+        max_result_per_keyword: usize,
+        max_depth: usize,
+        fetch_chains_batch_size: usize,
+        base_url: Option<String>,
+    ) -> PyResult<HashMap<KeywordPy, Vec<LocationPy>>> {
+        let mut findex = pyo3_unwrap!(FindexCloudRust::new(token, base_url), "error reading token");
+        let master_key = findex.token.findex_master_key.clone();
+
+        let keywords_set: HashSet<Keyword> =
+            keywords.iter().map(|keyword| keyword.0.clone()).collect();
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:312]: Same here

File: crates/findex/src/pyo3/py_api.rs [hunk 18]
```diff
@@ -554,3 +232,188 @@ impl InternalFindex {
...
+                    0,
+                )
+            ),
+            "error blocking for search"
+        );
+
+        Ok(search_results_to_python(results))
+    }
+
+    /// Generate a new Findex Cloud token with reduced permissions
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:338]: Missing parameter documentation
```suggestion
    /// Generates a new Findex Cloud token with reduced permissions.
```

File: crates/findex/src/pyo3/py_api.rs [hunk 19]
```diff
@@ -554,3 +232,188 @@ impl InternalFindex {
...
+
+        pyo3_unwrap!(
+            token.reduce_permissions(search, index),
+            "error reducing token permissions"
+        );
+
+        Ok(token.to_string())
+    }
+
+    /// Generate a new random Findex Cloud token
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_api.rs:351]: Missing parameter documentation
```suggestion
    /// Generates a new random Findex Cloud token.
```

File: crates/findex/src/pyo3/py_callbacks.rs [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+                .extract(py)
+                .map_err(|e| FindexPyo3Error::ConversionError(format!("{e} (fetch_entry)")))?;
+
+            // Convert python result (HashSet<[u8; UID_LENGTH]>) to
+            // HashSet<Uid<UID_LENGTH>>
+            let entry_table_items = py_result_table
+                .into_iter()
+                .map(Uid::from)
+                .collect::<HashSet<_>>();
+            Ok(entry_table_items)
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_callbacks.rs:95]: Style only
```suggestion
            Ok(py_result_table.into_iter().map(Uid::from).collect())
```

File: crates/findex/src/pyo3/py_callbacks.rs [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+        })
+    }
+
+    async fn fetch_all_entry_table_uids(
+        &self,
+    ) -> Result<HashSet<Uid<UID_LENGTH>>, FindexPyo3Error> {
+        Python::with_gil(|py| {
+            let results = self.fetch_all_entry_table_uids.call0(py).map_err(|e| {
+                FindexPyo3Error::Callback(format!("{e} (fetch_all_entry_table_uids)"))
+            })?;
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_callbacks.rs:84]: You may want to use the `pyo3_unwrap!()` marco (needs to implement `From<pyo3::PyErr>` for FindexPyo3Error):
```suggestion
            let results = pyo3_unwrap!(
                self.fetch_all_entry_table_uids.call0(py),
                "error fetching entry table"
            );
```

File: crates/findex/src/pyo3/py_callbacks.rs [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+                .map_err(|e| FindexPyo3Error::ConversionError(format!("{e} (fetch_entry)")))?;
+
+            // Convert python result (HashMap<[u8; UID_LENGTH], Vec<u8>>) to
+            // EncryptedEntryTable<UID_LENGTH>
+            let entry_table_items = py_result_table
+                .into_iter()
+                .map(|(k, v)| (Uid::from(k), v))
+                .collect::<HashMap<_, _>>();
+
+            Ok(entry_table_items.into())
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_callbacks.rs:123]: Style only (same everywhere):
```suggestion
            Ok(py_result_table
                .into_iter()
                .map(|(k, v)| (Uid::from(k), v))
                .collect())
```

File: crates/findex/src/pyo3/py_structs.rs [hunk 23]
```diff
@@ -1,106 +1,68 @@
...
-    ///     IndexedValue
-    #[staticmethod]
-    pub fn from_keyword(keyword_bytes: &[u8]) -> Self {
-        Self(IndexedValueRust::NextKeyword(Keyword::from(keyword_bytes)))
+fn truncate(s: String, max_chars: usize) -> String {
+    match s.char_indices().nth(max_chars) {
+        None => s,
+        Some((idx, _)) => format!("{}...", &s[..idx]),
     }
+}
```
Comment (tbrezot) [inline crates/findex/src/pyo3/py_structs.rs:16]: A function already exists in the standard implementation of `String` that does what you want but without copy (it drops terminating chars without reallocating):
```suggestion
fn truncate(s: String, max_chars: usize) -> String {
    s.truncate(max_chars);
    s
}
```

File: crates/findex/src/pyo3/mod.rs (hunk 14 above)
Comment (ackRow) [reply crates/findex/src/pyo3/mod.rs:53]: to match the functionalities of JS and Java libs.

File: crates/findex/python/cloudproof_findex/__init__.py (hunk 10 above)
Comment (ackRow) [reply crates/findex/python/cloudproof_findex/__init__.py:6]: it's a trick regarding python type checking for `List` of union types and type inferences: https://mypy.readthedocs.io/en/latest/common_issues.html#variance

File: crates/findex/src/pyo3/py_callbacks.rs (hunk 21 above)
Comment (ackRow) [reply crates/findex/src/pyo3/py_callbacks.rs:84]: There is two types of exceptions to raise: `FindexPyo3Error::Callback` and `FindexPyo3Error::ConversionError`.
This is why we did not implement `From<pyo3::PyErr>` before.

File: crates/findex/src/pyo3/py_structs.rs (hunk 23 above)
Comment (tbrezot) [reply crates/findex/src/pyo3/py_structs.rs:16]: You could have truncated to `max_char + 3` then rewrite the last three chars by `.`

**Pull request merged**